jplag.frontend.rust/src/test/java/de/jplag/rust/RustFrontendTest.java

```diff
@@ -0,0 +1,148 @@
+     */
+    private static final String RUST_EMPTY_OR_SINGLE_LINE_COMMENT = "\\s*(?://.*)?";
+    private static final String RUST_MULTILINE_COMMENT_BEGIN = "\\s*/\\*.*";
+    private static final String RUST_MULTILINE_COMMENT_END = ".*\\*/\\s*";
+
+    /**
+     * Test source file that is supposed to produce a complete set of tokens, i.e. all types of tokens.
+     */
+    private static final String COMPLETE_TEST_FILE = "complete.rs";
+    public static final int NOT_SET = -1;
+    private static final String EMPTY_STRING = "";
+    private static final String RUST_SHEBANG = "#!.*$";
+    private static final double EPSILON = 1E-6;
+
+    private final Logger logger = LoggerFactory.getLogger("Rust frontend test");
+    private final String[] testFiles = new String[] {"deno_core_runtime.rs", COMPLETE_TEST_FILE};
+    private final File testFileLocation = Path.of("src", "test", "resources", "de", "jplag", "rust").toFile();
+    private Language language;
+
+    @BeforeEach
+    void setup() {
+        language = new Language();
+    }
+
+    @Test
+    void parseTestFiles() {
+        for (String fileName : testFiles) {
+            TokenList tokens = language.parse(testFileLocation, new String[] {fileName});
+            String output = TokenPrinter.printTokens(tokens, testFileLocation, List.of(fileName));
+            logger.info(output);
+
+            testSourceCoverage(fileName, tokens);
+            if (fileName.equals(COMPLETE_TEST_FILE))
+                testTokenCoverage(tokens, fileName);
+        }
+    }
+
+    /**
+     * Confirms that the code is covered to a basic extent, i.e. each line of code contains at least one token.
+     * @param fileName a code sample file name
+     * @param tokens the TokenList generated from the sample
+     */
+    private void testSourceCoverage(String fileName, TokenList tokens) {
+        File testFile = new File(testFileLocation, fileName);
+
+        try {
+            List<String> lines = Files.readAllLines(testFile.toPath());
+
+            // All lines that contain code
+            var codeLines = new ArrayList<>(getCodeLines(lines));
+            // All lines that contain token
+            var tokenLines = IntStream.range(0, tokens.size()).mapToObj(tokens::getToken).mapToInt(Token::getLine).distinct().boxed().toList();
+
+            // Keep only lines that have no tokens
+            codeLines.removeAll(tokenLines);
+
+            double coverage = 1.d - (codeLines.size() * 1.d / (codeLines.size() + tokenLines.size()));
+            if (coverage == 1) {
+                logger.info("All lines covered.");
+            } else {
+                logger.info("Coverage: %.1f%%.".formatted(coverage * 100));
+                logger.info("Missing lines {}", codeLines);
+                if (coverage - 0.9 <= EPSILON) {
+                    // TODO use fail() instead when frontend is ready
+                    logger.error("Source coverage is unsatisfactory");
+                }
+            }
```

Comment:
If in the future something about the frontend changes, it will be almost impossible to detect what changed and why the coverage is now reduced. Thus, I would suggest to explicitly document the lines of the code that are not covered. One option might be to indicate block-wise uncovered sections with surrounding comment (like `//TOKENIZATION IGNORED SECTION START` / `//TOKENIZATION IGNORED SECTION END`) or for single lines a trailing comment.

Reply:
The files used for the test are very long, so it would be hard to annotate it that way. 
Maybe I can add the output of the TokenPrinter for the test files somewhere so a diff analysis tool can reveal what changed.
Would that be a fair option? -- If yes, I need a hint where to put those output files.

Reply:
Ah, I'm sorry. I confused it with the token coverage test where only the `complete.rs` file is relevant. Yes, for the other file it is probably too much.
Your proposed solution sounds like a good compromise. Maybe put it in `resources/de/jplag/rust/annotated`?